test/functional/feature_utxo_set_hash.py

```diff
@@ -0,0 +1,90 @@
+#!/usr/bin/env python3
+# Copyright (c) 2020 The Bitcoin Core developers
+# Distributed under the MIT software license, see the accompanying
+# file COPYING or http://www.opensource.org/licenses/mit-license.php.
+"""Test UTXO set hash value calculation.
+
+Based on the UTXO set of a P2P connection serialize and calculate a hash
+for the UTXO set and compare it with the actual node's hash.
+"""
+
+from test_framework.blocktools import (
+    create_block,
+    create_coinbase,
+    create_tx_with_script,
+)
+from test_framework.messages import (
+    COIN,
+)
+from test_framework.p2p import P2PDataStore
+from test_framework.muhash import MuHash3072
+from test_framework.test_framework import BitcoinTestFramework
+from test_framework.util import assert_equal
+
+import hashlib
+import struct
+
+class UTXOSetHashTest(BitcoinTestFramework):
+    def set_test_params(self):
+        self.num_nodes = 1
+        self.setup_clean_chain = True
+
+    def run_test(self):
+        node = self.nodes[0]
+        conn = node.add_p2p_connection(P2PDataStore())
+
+        # Get tip information (Genesis) to build the next block.
+        best_block_hash = node.getbestblockhash()
+        tip = int(best_block_hash, 16)
+        best_block = node.getblock(best_block_hash)
+        height = best_block["height"] + 1
+        block_time = best_block["time"] + 1
+
+        # Generate 100 blocks so the first coinbase is matured. We have to do
+        # it in a way that conn knows about them so we use
+        # send_blocks_and_test.
+        for i in range(100):
+            # Coinbases are anyone can spend.
+            block = create_block(tip, create_coinbase(height), block_time)
+            block.solve()
+            conn.send_blocks_and_test([block], node, success=True)
+
+            tip = block.sha256
+            best_block = block
+            height += 1
+            block_time += 1
+
+            # Save first block because we want to spend its coinbase.
+            if i == 0:
+                block1 = block
+
+        # Generate another block with one tx spending a coinbase and another
+        # spending that tx's output.
+        tx1 = create_tx_with_script(block1.vtx[0], 0, script_sig=b'\x51', amount=25 * COIN)
+        tx2 = create_tx_with_script(tx1, 0, script_sig=b'\x51', amount=25 * COIN)
+
+        block2 = create_block(tip, create_coinbase(height), block_time)
+        block2.vtx.extend([tx1, tx2])
+        block2.hashMerkleRoot = block2.calc_merkle_root()
+        block2.rehash()
+        block2.solve()
+
+        conn.send_blocks_and_test([block2], node, success=True)
+
+        # Calculate the MuHash of the UTXO set estimate
+        muhash = MuHash3072()
```

Comment:
Would this test be a little better if instead of forming the utxo set at the end, you maintained a running sum and added as blocks were mined, and then added/subtracted when UTXOs were spent and created in `tx1`, `tx2`?

Reply:
I don't think so because we need to somehow keep track of what is really in the UTXO set and I think this would make test harder to follow if I did it this way. And I feel like this is more of the standard way of how we structure our tests.